x-pack/apm-server/sampling/eventstorage/storage_manager.go

```diff
@@ -341,30 +397,64 @@ func (sm *StorageManager) WriteSubscriberPosition(data []byte) error {
 // For testing only.
 func (sm *StorageManager) NewUnlimitedReadWriter() StorageLimitReadWriter {
-	return sm.NewReadWriter(0)
+	return sm.NewReadWriter(0, 0)
 }
 
-// NewReadWriter returns a read writer with storage limit.
-func (sm *StorageManager) NewReadWriter(storageLimit uint64) StorageLimitReadWriter {
+// NewReadWriter returns a read writer configured with storage limit and disk threshold ratio.
+func (sm *StorageManager) NewReadWriter(storageLimit uint64, diskThresholdRatio float64) StorageLimitReadWriter {
 	splitRW := SplitReadWriter{
 		eventRW:    sm.eventStorage.NewReadWriter(),
 		decisionRW: sm.decisionStorage.NewReadWriter(),
 	}
 
-	dbStorageLimit := func() uint64 {
-		return storageLimit
-	}
-	if storageLimit == 0 {
-		sm.logger.Infof("setting database storage limit to unlimited")
+	// dbStorageLimit returns max size of db in bytes.
+	// If size of db exceeds dbStorageLimit, writes should be rejected.
+	var dbStorageLimit func() uint64
+	if sm.diskStatFailed.Load() && storageLimit == 0 && diskThresholdRatio > 0 {
+		dbStorageLimit = func() uint64 {
+			return dbStorageLimitFallback
+		}
+		sm.logger.Warnf("failed to get disk usage; overriding database storage limit to fallback default of %0.1fgb", float64(dbStorageLimitFallback)/gb)
 	} else {
-		sm.logger.Infof("setting database storage limit to %.1fgb", float64(storageLimit))
+		dbStorageLimit = func() uint64 {
+			return storageLimit
+		}
+		if storageLimit == 0 {
+			sm.logger.Infof("setting database storage limit to unlimited")
+		} else {
+			sm.logger.Infof("setting database storage limit to %0.1fgb", float64(storageLimit)/gb)
+		}
 	}
 
 	// To limit db size to storage_limit
 	dbStorageLimitChecker := NewStorageLimitCheckerFunc(sm.dbSize, dbStorageLimit)
 	dbStorageLimitRW := NewStorageLimitReadWriter("database storage limit", dbStorageLimitChecker, splitRW)
 
-	return dbStorageLimitRW
+	// diskThreshold returns max used disk space in bytes.
+	// If size of used disk space exceeds diskThreshold, writes should be rejected.
+	var diskThreshold func() uint64
+	if sm.diskStatFailed.Load() {
+		diskThreshold = func() uint64 {
+			return 0 // unlimited
+		}
+		sm.logger.Warnf("failed to get disk usage; disabling disk threshold check")
+	} else {
+		diskThreshold = func() uint64 {
+			return uint64(float64(sm.diskStat.total.Load()) * diskThresholdRatio)
+		}
+		// the total disk space could change in runtime, but it is still useful to print it out in logs.
+		sm.logger.Infof("setting disk threshold ratio to %.2f of total disk space of %0.1fgb", diskThresholdRatio, float64(sm.diskStat.total.Load())/gb)
+	}
+
+	// To limit actual disk usage percentage to diskThresholdRatio
+	diskThresholdChecker := NewStorageLimitCheckerFunc(sm.diskUsed, diskThreshold)
+	diskThresholdRW := NewStorageLimitReadWriter(
+		fmt.Sprintf("disk usage ratio exceeding threshold of %.2f", diskThresholdRatio),
+		diskThresholdChecker,
+		dbStorageLimitRW,
+	)
```

Comment:
This gets fairly complex now. I think it is fine to move forward given the time constraints, but would appreciate a bit of refactoring soon after. The responsibilities of ensuring disk space is available is now spread amongst several parts.